### ISSUE TITLE
Imagenet training example missing learning rate decay?

### ISSUE THREAD

**Issue by yanii** [chainer/chainer#1549]
As far as I can tell there doesn't seem to be any change in learning rate for the imagenet example, or for that matter in any of the examples. Am I missing something?

Before the Trainer abstraction (c67fa7869071c6470fd666be4be7f3c021d507d2), this line in the training loop decayed learning rate:
`optimizer.lr *= 0.97`

From what I can tell, the ExponentialShift extension should be used in the Trainer abstraction to perform a similar function, but there is none used. In fact there are no examples whatsoever on how to decrease learning rate during training that I can find. It looks like this might be correct however:

`trainer.extend(extensions.ExponentialShift('lr', 0.97), trigger=(1, epoch))`

How to print the current learning rate with a reporter is something I can't figure out within this framework. I suppose you could put a reporter in the ExponentialShift/Optimizer?

I'm personally not a fan of the abstraction to begin with, but if you are going to add this stuff, the examples should be fully fleshed out and easy to understand for newcomers. I've resorted to going back in the history and using the old examples more than a few times now, they were far clearer and more flexible (which frankly is what we need for research purposes - not an inordinate number of layers of abstraction). The harsh truth is that if I was coming across chainer for the first time now, with only these examples to look at, I'd probably have never ended up using it.

**Comment by delta2323** [comment]
Thank you for your comments and sorry for our late response.

As you pointed out, we accidentally dropped the learning rate decay from the imagenet example. As this is a bug of the example, we should restore it.

As for the reporting of the learning rate, we can use `observe_lr` extension, which is implemented in PR #1472. You can check its typical usage in `examples/imagenet/train_imagenet.py`.

**Comment by beam2d** [comment]
Thank you for giving us your opinion on the current direction towards training loop abstraction. Our motivation is to avoid boilerplate coding on writing training loop and support frequently-used but heavy-to-support features around managing the training process. We really do not want to hinder users' attempts to write their own training loop, but I believe that at least for beginners it would be better to use the prepared training loop instead of their own, because it makes them much easier to avoid a bug and can easily support frequently-used features (the most important one that I expected is suspend/resume of the training process with properly managing all stuffs involved in the training process and its monitoring).

Anyway, we keep the design support user's own training loop, thus the training loop abstraction is just built on top of other stuffs and users can use the underlying features without the trainer, etc. Sorry for removing the code examples for writing one's own training loop; I will consider recover such examples to the official repository so that users can easily find the way to write a training loop without using the trainer.

**Comment by yanii** [comment]
@delta2323 Thanks for confirming this. I think the potentially bigger issue this raises is that these examples obviously haven't been tested for a while. There is no way that imagenet example could get close to any reasonable accuracy without the LR change.

@beam2d Thanks for spending the time to reply to my concerns, it's much appreciated. I think re-instating the original code examples (as 'custom training loop' examples or something alongside the abstracted versions) would go a long way towards this, because without these it is really not clear how flexible and easy Chainer is as a general neural net API, and for me at least that is by far it's biggest strength. It's also no longer obvious how to do this from scratch.

As you point out, flexibility is good, and perhaps the training loops help beginners. I'd be willing to bet the majority of users of Chainer are actually using it rather than say Caffe precisely because of the flexibility it allows, rather than just another framework to push images through a model for inference though (where frankly it's hard to compete), and it would be good to keep this niche in mind. Perhaps I'm biased towards my own usage though :-)
